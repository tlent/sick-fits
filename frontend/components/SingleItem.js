import React, { Component } from "react";
import gql from "graphql-tag";
import { Query } from "react-apollo";
import styled from "styled-components";
import Head from "next/head";
import Error from "./ErrorMessage";

const SingleItemStyles = styled.div`
  max-width: 1200px;
  margin: 2rem auto;
  box-shadow: ${props => props.theme.bs};
  display: grid;
  grid-auto-columns: 1fr;
  grid-auto-flow: column;
  min-height: 800px;

  img {
    width: 100%;
    height: 100%;
    object-fit: contain;
  }

  .details {
    margin: 3rem;
    font-size: 2rem;
  }
`;

const SINGLE_ITEM_QUERY = gql`
  query SINGLE_ITEM_QUERY($id: ID!) {
    item(where: { id: $id }) {
      id
      title
      description
      image
      largeImage
    }
  }
`;

class SingleItem extends Component {
  render() {
    const { id } = this.props;
    return (
      <Query query={SINGLE_ITEM_QUERY} variables={{ id }}>
        {({ error, loading, data }) => {
          if (error) return <Error error={error} />;
          if (loading) return <p>Loading!</p>;
          const { item } = data;
          if (!item) return <p>No item found for {id}</p>;
          return (
            <SingleItemStyles>
              <Head>
                <title>Sick Fits | {item.title}</title>
              </Head>
              <img src={item.largeImage} alt={item.title} />
              <div className="details">
                <h2>Viewing {item.title}</h2>
                <p>{item.description}</p>
              </div>
            </SingleItemStyles>
          );
        }}
      </Query>
    );
  }
}

export default SingleItem;
