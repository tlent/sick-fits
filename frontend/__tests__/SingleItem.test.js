import { mount } from "enzyme";
import wait from "waait";
import SingleItem, { SINGLE_ITEM_QUERY } from "../components/SingleItem";
import { MockedProvider } from "react-apollo/test-utils";
import { fakeItem } from "../lib/testUtils";

describe("SingleItem component", () => {
  it("renders with proper data", async () => {
    const mocks = [
      {
        request: { query: SINGLE_ITEM_QUERY, variables: { id: "123" } },
        result: { data: { item: fakeItem() } }
      }
    ];
    const wrapper = mount(
      <MockedProvider mocks={mocks}>
        <SingleItem id={"123"} />
      </MockedProvider>
    );
    expect(wrapper.text()).toContain("Loading");
    await wait();
    wrapper.update();
    expect(wrapper.find("h2")).toMatchSnapshot();
    expect(wrapper.find("img")).toMatchSnapshot();
    expect(wrapper.find("p")).toMatchSnapshot();
  });

  it("errors with a not found item", async () => {
    const mocks = [
      {
        request: { query: SINGLE_ITEM_QUERY, variables: { id: "123" } },
        result: { errors: [{ message: "Items not found!" }] }
      }
    ];
    const wrapper = mount(
      <MockedProvider mocks={mocks}>
        <SingleItem id={"123"} />
      </MockedProvider>
    );
    await wait();
    wrapper.update();
    const item = wrapper.find('[data-test="graphql-error"]');
    expect(item.text()).toContain("Items not found!");
    expect(item).toMatchSnapshot();
  });
});
