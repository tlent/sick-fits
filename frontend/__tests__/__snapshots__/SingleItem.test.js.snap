// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SingleItem component errors with a not found item 1`] = `
<p
  data-test="graphql-error"
>
  <strong>
    Shoot!
  </strong>
  Items not found!
</p>
`;

exports[`SingleItem component renders with proper data 1`] = `
<h2>
  Viewing 
  dogs are best
</h2>
`;

exports[`SingleItem component renders with proper data 2`] = `
<img
  alt="dogs are best"
  src="dog.jpg"
/>
`;

exports[`SingleItem component renders with proper data 3`] = `
<p>
  dogs
</p>
`;
